iwyu.cc

```diff
@@ -1761,6 +1761,20 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     return true;
   }
 
+  bool TraverseCXXForRangeStmt(CXXForRangeStmt* stmt) {
```

Comment:
It might be nice to add a short description as to why this is necessary (from the commit message) as a code comment here.

And number two; I wonder if there's a way to detect at reporting time that the use is in implicit code and discard it there instead? I guess I don't 100% understand what the AST looks like in this case.

Reply:
I'm not sure that inspecting AST node stack in order to search `CXXForRangeStmt` node from reporting visitors directly would be simpler (besides that it could affect performance). But the following works:
```cpp
bool TraverseDeclStmt(clang::DeclStmt* stmt) {
  if (const auto* range_for =
          current_ast_node()->template GetParentAs<CXXForRangeStmt>()) {
    if (stmt->getSingleDecl()->isImplicit() &&
        range_for->getRangeStmt() != stmt) {
      return true;
    }
  }
  return Base::TraverseDeclStmt(stmt);
}
```

Reply:
> I guess I don't 100% understand what the AST looks like in this case.

The code
```cpp
for (char ch; int x : container)
  float f;
```
is equivalent to
```cpp
char ch;
auto && __range1 = container;
auto __begin1 = __range1.begin();
auto __end1 = __range1.end();
for ( ; __begin1 != __end1; ++__begin1) {
  int x = *__begin1;
  float f;
}
```
hence produces such an AST:
![изображение](https://github.com/include-what-you-use/include-what-you-use/assets/32954549/6a558e0e-a571-43fc-9f7a-603156d75200)

The problem arises on reporting type of base expression of `CXXMemberCallExpr`.

Note that "InitExpr" still needs to be traversed because it is explicitly written in the source despite it initializes the `__range1` implicit variable.

Reply:
In fact, the iterator type info and iterator operator declarations are also needed, of course, but it is rare situation when they are not provided by the container class header, and reporting them may just produce unwanted noise in the output.

Reply:
Yes, I thought about it as well, but it breaks at least this case:
```cpp
void TakeByRef(const Indirect&);

void TestImplicit(Indirect& a){
  int array[] = {1,2,3};
  for (int x : array) {
    [=] {
        TakeByRef(a);
    }();
  }
}
```
The complete `Indirect` type info is needed due to the _implicit_ capture by value.

Reply:
Another random thought I had was that it would be cool if we could build an incremental resugar-map as we traverse the normal tree too, not just for templates. So e.g. `TraverseCXXForRangeStmt` could extract all types from all its immediate children and map desugared to whatever sugared is available. Then when we visit the children we could simply try-resugar from that map.

Reply:
That's interesting idea...

However, imagine that `begin()` and `end()` methods of some container class of some library return custom `Iterator` type. That type may even not be aliased, but the container-defining header almost definitely would "provide" it by including the corresponding header. The custom traversal of `CXXForRangeStmt` node could allow to decide whether report it or not, based on its "provision status".

Reply:
Makes sense, thanks!